# 15) Veri Seti Tanıyalım

Bir analize başlamadan önce, veri setini sistematik olarak tanımak gerekir aksi halde ilerleyen aşamalarda (eksik değer, aykırı değer, hipotez 
testi) sürprizlerle karşılaşılabilir. Ek olarak, bir veri setini analize hazır hale getirmek için veri setindeki olası sorunları (boş değer, aykırı değer, yanlış veri tipi, normal dağılımın olmaması ...) çözersek ileriki analizlerimizde daha iyi sonuçlar alabiliriz. Bu tanıma süreci genelde şu adımlardan oluşur:

1. **Boyut:** Kaç satır, kaç kolon var? (`.shape`)
2. **Veri Tipleri:** Her kolon hangi tipte kayıtlı? (`.dtypes`) — bu, 
   Konu 3'te gördüğümüz nicel/nitel ayrımını teyit etmemizi sağlar.
3. **Eksik Değer Kontrolü:** Herhangi bir kolonda boş/eksik veri var mı? 
   (`.isnull().sum()`) — çözümünü ileride (Eksik Değer bölümünde) 
   işleyeceğiz, burada sadece "var mı yok mu" tespiti yapıyoruz.
4. **Betimsel Özet:** Sayısal kolonlar için `.describe()` — ortalama, std, 
   min, max, kartiller tek seferde görülür.
5. **Kategorik Kolonların Benzersiz Değerleri:** `.unique()` / `.nunique()` 
   — bir kategorik değişkenin kaç farklı grup içerdiğini gösterir.

Bu 5 adım, bir veri setiyle çalışmaya başlamadan önce atılması gereken 
**ilk keşif (initial exploration)** adımlarıdır.

Örnek olması adına Seaborn kütüphanesinden exercise veri setini inceleyelim.
Ek Bilgi: Pandas, Python'un veriyi analize hazır hale getirme sürecinde en çok kullanılan ve veriyi keşfetmemizi sağlayan açık kaynaklı bir kütüphanedir. Biz de bu kütüphanenin fonksiyonlarından yararlanacağız.

In [15]:
import seaborn as sns
import pandas as pd
df = sns.load_dataset("exercise")

In [16]:
# excel'deki tablo yapısı Pandas'ta DataFrame olarak adlandırılır. DataFrame, satır ve sütunlardan oluşan bir veri yapısıdır.
print(f"Veri setinin gözlem (satır) sayısı: {df.shape[0]}")
print(f"Veri setinin sütun sayısı: {df.shape[1]}")

Veri setinin gözlem (satır) sayısı: 90
Veri setinin sütun sayısı: 6


In [17]:
print(df.dtypes) # Veri setindeki her bir sütunun veri tipini öğrenmek için dtypes özelliğini kullanabiliriz.

Unnamed: 0       int64
id               int64
diet          category
pulse            int64
time          category
kind          category
dtype: object


In [18]:
# Sütun Bazlı Boş Değer Kontrolü
print(df.isnull().sum())
print()
print(df.isna().sum().sum()) # isna() ve isnull() fonksiyonları aynı işlemi yapar. Her iki fonksiyon da veri setindeki boş değerleri kontrol etmek için kullanılır. 2 kere toplam almak verinin tamamındaki boş hücre sayısını verir.


Unnamed: 0    0
id            0
diet          0
pulse         0
time          0
kind          0
dtype: int64

0


In [19]:
print("Betimsel İstatistikler")
print(df.describe())

Betimsel İstatistikler
       Unnamed: 0         id       pulse
count   90.000000  90.000000   90.000000
mean    44.500000  15.500000   99.700000
std     26.124701   8.703932   14.858471
min      0.000000   1.000000   80.000000
25%     22.250000   8.000000   90.250000
50%     44.500000  15.500000   96.000000
75%     66.750000  23.000000  103.000000
max     89.000000  30.000000  150.000000


In [20]:
# Veri setindeki sadece kategorik degiskenleri cekelim.
categorical = df.select_dtypes(include=["category"]) # Belirli bir değişkene ait sütunları filtrelemek için select_dtypes() fonksiyonunu kullanabiliriz. include parametresi ile hangi veri tiplerini seçmek istediğimizi belirtebiliriz. Burada "category" veri tipine sahip sütunları seçiyoruz.
print(categorical)

       diet    time     kind
0   low fat   1 min     rest
1   low fat  15 min     rest
2   low fat  30 min     rest
3   low fat   1 min     rest
4   low fat  15 min     rest
..      ...     ...      ...
85   no fat  15 min  running
86   no fat  30 min  running
87   no fat   1 min  running
88   no fat  15 min  running
89   no fat  30 min  running

[90 rows x 3 columns]


In [21]:
print(df["diet"].unique()) # unique() fonksiyonu, bir sütundaki benzersiz değerleri döndürür.
print(df["diet"].nunique()) # nunique() fonksiyonu, bir sütundaki benzersiz değerlerin sayısını döndürür.

['low fat', 'no fat']
Categories (2, object): ['no fat', 'low fat']
2


In [22]:
print("*****Veri Seti Hakkındaki Genel Bilgiler*****")
df.info() # info() fonksiyonu, veri seti hakkında genel bilgiler verir. Bu bilgiler arasında sütun isimleri, veri tipleri, boş değer sayısı ve bellek kullanımı gibi bilgiler bulunur.

*****Veri Seti Hakkındaki Genel Bilgiler*****
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90 entries, 0 to 89
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   Unnamed: 0  90 non-null     int64   
 1   id          90 non-null     int64   
 2   diet        90 non-null     category
 3   pulse       90 non-null     int64   
 4   time        90 non-null     category
 5   kind        90 non-null     category
dtypes: category(3), int64(3)
memory usage: 2.9 KB


## Sonuç
Veri setimiz 90 gözlem, 6 sütundan oluşuyor. Hiç bir sütunda boş değer bulunmamaktadır. Veri setindeki kind, time, diet sütunları kategori veri tipindedir. Diet sütununda 'low fat' ve 'no fat' olmak üzere 2 farklı kategori vardır. Veri seti hafızada 2.9 KiloBayt yer kaplamaktadır.